packages/template-retail-react-app/app/utils/jwt-utils.js

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {createRemoteJWKSet as joseCreateRemoteJWKSet, jwtVerify} from 'jose'
+import {getAppOrigin} from '@salesforce/pwa-kit-react-sdk/utils/url'
+import {getConfig} from '@salesforce/pwa-kit-runtime/utils/ssr-config'
+
+const throwSlasTokenValidationError = (message, code) => {
+    throw new Error(`SLAS Token Validation Error: ${message}`, code)
+}
+
+export const createRemoteJWKSet = () => {
+    const appOrigin = getAppOrigin()
+    const {app: appConfig} = getConfig()
+    const shortCode = appConfig.commerceAPI.parameters.shortCode
+    const tenantId = appConfig.commerceAPI.parameters.organizationId.replace(/^f_ecom_/, '')
+    const JWKS_URI = `${appOrigin}/${shortCode}/${tenantId}/oauth2/jwks`
+    return joseCreateRemoteJWKSet(new URL(JWKS_URI))
+}
+
+export const validateSlasCallbackToken = async (token) => {
+    try {
+        const jwks = createRemoteJWKSet()
+        const {payload} = await jwtVerify(token, jwks, {})
+        return payload
+    } catch (error) {
+        throwSlasTokenValidationError(error.message, 401)
+    }
+}
+
+const tenantIdRegExp = /^[a-zA-Z]{4}_([0-9]{3}|s[0-9]{2}|stg|dev|prd)$/
+const shortCodeRegExp = /^[a-zA-Z0-9-]+$/
+
+/**
+ *  Handles JWKS (JSON Web Key Set) caching the JWKS response for 2 weeks.
+ *
+ * @param {object} req Express request object.
+ * @param {object} res Express response object.
+ * @param {object} options Options for fetching B2C Commerce API JWKS.
+ * @param {string} options.shortCode - The Short Code assigned to the realm.
+ * @param {string} options.tenantId - The Tenant ID for the ECOM instance.
+ * @returns {Promise<*>} Promise with the JWKS data.
+ */
+export async function jwksCaching(req, res, options) {
+    const {shortCode, tenantId} = options
+
+    const isValidRequest = tenantIdRegExp.test(tenantId) && shortCodeRegExp.test(shortCode)
+    if (!isValidRequest)
+        return res
+            .status(400)
+            .json({error: 'Bad request parameters: Tenant ID or short code is invalid.'})
+    try {
+        const JWKS_URI = `https://${shortCode}.api.commercecloud.salesforce.com/shopper/auth/v1/organizations/f_ecom_${tenantId}/oauth2/jwks`
+        const response = await fetch(JWKS_URI, {
+            headers: {
+                'User-Agent': 'OctoperfMercuryPerfTest'
+            }
```

Comment:
TODO: Remove before merging into feature branch.

@adamraya When testing with staging, this header was required to bypass some Cloudflare restrictions. Was this something you saw when testing with Storefront Preview as well?

Reply:
No. At the time we implemented this feature in the private BFF layer, we did not need to add any header.

Reply:
Confirmed with Pratik that customer ECOM instances should hit production CDN so there should be no Cloudflare WAFs blocking this API call!